### ISSUE TITLE
M4 — Module system and the core layers

### ISSUE THREAD

**Issue by harshitagrawal2O** [harshitagrawal2O/tensorkit#4]
**Milestone 4 of 8** — see [`SPEC.md`](../blob/main/SPEC.md) section 4 for the full definition of done.

**CS subject:** OOP (composition, parameter registration)

**Nobody is working on this.** Comment to claim it. One person per milestone at a time; if a
claim goes quiet for two weeks it is fair game again.

## Scope

`Module` base with recursive `parameters()`, `state_dict()`, `train()`/`eval()`. Then `Linear`, `ReLU`, `GELU`, `Softmax`, `Sequential`, `Flatten`, `Dropout`, `Embedding`.

## Definition of done

- Weight tying yields each parameter **once** from `parameters()` (I-MOD-UNIQUE)
- `Dropout` in `eval()` is bitwise the identity; in `train()` the output expectation equals the input
- `Embedding` backward scatter-**adds** over repeated indices

## Where it gets interesting

The embedding backward. `grad[idx] = v` keeps the last write; indices repeat constantly in a batch, so every frequent token would learn from exactly one occurrence.

## Ground rules

- Do **not** weaken a test to make it pass. If you believe a test is wrong, open a
  [spec-defect issue](../issues/new?template=spec-defect.yml) instead — several tests have
  already been corrected that way.
- Keep the existing docstrings; replace only the `raise NotImplementedError(...)` line.
- `ruff check`, `ruff format --check` and `mypy` must be clean.

Chapter-length background is in [the book](../blob/main/docs/tensorkit-book.pdf).

```bash
python -m pytest -m m4 -q
```
